In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table'),
 ('countries', 'view'),
 ('purchased_tracks', 'view'),
 ('all_tracks', 'view')]

In [3]:
%%sql
select * from employee

Done.


[(1, 'Adams', 'Andrew', 'General Manager', None, '1962-02-18 00:00:00', '2016-08-14 00:00:00', '11120 Jasper Ave NW', 'Edmonton', 'AB', 'Canada', 'T5K 2N1', '+1 (780) 428-9482', '+1 (780) 428-3457', 'andrew@chinookcorp.com'),
 (2, 'Edwards', 'Nancy', 'Sales Manager', 1, '1958-12-08 00:00:00', '2016-05-01 00:00:00', '825 8 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 2T3', '+1 (403) 262-3443', '+1 (403) 262-3322', 'nancy@chinookcorp.com'),
 (3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com'),
 (6, 'Mitchell', 'Michael', 'IT Manager', 1, '1973-07-01 00:00:00', '2016-10-17 00:00:00', '5827 Bowness Road NW', 'Calgary', 'AB', 'Canada', 'T3B 0C5', '+1 (403) 246-9887', '+1 (403) 246-9899', 'michael@chinookcorp.com'),
 (7, 'King', 'Robert', 'IT Staff', 6, '1970-05-29 00:00:00', '2017-01-02 00:00:00', '590 Columbia Boulevard West', 'Lethbridge', 'AB', 'Canada', 'T1K 5N8', '+1 (403) 456-9986', '+1 (403) 456-8485', 'robert@chinookcorp.com'),
 (8, 'Callahan', 'Laura', 'IT Staff', 6, '1968-01-09 00:00:00', '2017-03-04 00:00:00', '923 7 ST NW', 'Lethbridge', 'AB', 'Canada', 'T1H 1Y8', '+1 (403) 467-3351', '+1 (403) 467-8772', 'laura@chinookcorp.com')]

In [4]:
%%sql
select * from playlist inner join playlist_track on playlist.playlist_id=playlist_track.playlist_id limit 100

Done.


[(1, 'Music', 1, 3402),
 (1, 'Music', 1, 3389),
 (1, 'Music', 1, 3390),
 (1, 'Music', 1, 3391),
 (1, 'Music', 1, 3392),
 (1, 'Music', 1, 3393),
 (1, 'Music', 1, 3394),
 (1, 'Music', 1, 3395),
 (1, 'Music', 1, 3396),
 (1, 'Music', 1, 3397),
 (1, 'Music', 1, 3398),
 (1, 'Music', 1, 3399),
 (1, 'Music', 1, 3400),
 (1, 'Music', 1, 3401),
 (1, 'Music', 1, 3336),
 (1, 'Music', 1, 3478),
 (1, 'Music', 1, 3375),
 (1, 'Music', 1, 3376),
 (1, 'Music', 1, 3377),
 (1, 'Music', 1, 3378),
 (1, 'Music', 1, 3379),
 (1, 'Music', 1, 3380),
 (1, 'Music', 1, 3381),
 (1, 'Music', 1, 3382),
 (1, 'Music', 1, 3383),
 (1, 'Music', 1, 3384),
 (1, 'Music', 1, 3385),
 (1, 'Music', 1, 3386),
 (1, 'Music', 1, 3387),
 (1, 'Music', 1, 3388),
 (1, 'Music', 1, 3365),
 (1, 'Music', 1, 3366),
 (1, 'Music', 1, 3367),
 (1, 'Music', 1, 3368),
 (1, 'Music', 1, 3369),
 (1, 'Music', 1, 3370),
 (1, 'Music', 1, 3371),
 (1, 'Music', 1, 3372),
 (1, 'Music', 1, 3373),
 (1, 'Music', 1, 3374),
 (1, 'Music', 1, 99),
 (1, 'Music', 1, 100),
 (1, 'Music', 1, 101),
 (1, 'Music', 1, 102),
 (1, 'Music', 1, 103),
 (1, 'Music', 1, 104),
 (1, 'Music', 1, 105),
 (1, 'Music', 1, 106),
 (1, 'Music', 1, 107),
 (1, 'Music', 1, 108),
 (1, 'Music', 1, 109),
 (1, 'Music', 1, 110),
 (1, 'Music', 1, 166),
 (1, 'Music', 1, 167),
 (1, 'Music', 1, 168),
 (1, 'Music', 1, 169),
 (1, 'Music', 1, 170),
 (1, 'Music', 1, 171),
 (1, 'Music', 1, 172),
 (1, 'Music', 1, 173),
 (1, 'Music', 1, 174),
 (1, 'Music', 1, 175),
 (1, 'Music', 1, 176),
 (1, 'Music', 1, 177),
 (1, 'Music', 1, 178),
 (1, 'Music', 1, 179),
 (1, 'Music', 1, 180),
 (1, 'Music', 1, 181),
 (1, 'Music', 1, 182),
 (1, 'Music', 1, 2591),
 (1, 'Music', 1, 2592),
 (1, 'Music', 1, 2593),
 (1, 'Music', 1, 2594),
 (1, 'Music', 1, 2595),
 (1, 'Music', 1, 2596),
 (1, 'Music', 1, 2597),
 (1, 'Music', 1, 2598),
 (1, 'Music', 1, 2599),
 (1, 'Music', 1, 2600),
 (1, 'Music', 1, 2601),
 (1, 'Music', 1, 2602),
 (1, 'Music', 1, 2603),
 (1, 'Music', 1, 2604),
 (1, 'Music', 1, 2605),
 (1, 'Music', 1, 2606),
 (1, 'Music', 1, 2607),
 (1, 'Music', 1, 2608),
 (1, 'Music', 1, 923),
 (1, 'Music', 1, 924),
 (1, 'Music', 1, 925),
 (1, 'Music', 1, 926),
 (1, 'Music', 1, 927),
 (1, 'Music', 1, 928),
 (1, 'Music', 1, 929),
 (1, 'Music', 1, 930),
 (1, 'Music', 1, 931),
 (1, 'Music', 1, 932),
 (1, 'Music', 1, 933),
 (1, 'Music', 1, 934),
 (1, 'Music', 1, 935)]

In [5]:
%%sql
with info as 
(select count(quantity) as number_tracks_sold ,genre.name as genre_name from invoice 
inner join invoice_line on invoice.invoice_id=invoice_line.invoice_id 
inner join track on invoice_line.track_id=track.track_id 
inner join genre on genre.genre_id=track.genre_id
where invoice.billing_country= "USA"
group by genre.genre_id
order by number_tracks_sold desc)


select number_tracks_sold, 
round(cast(number_tracks_sold as float)/(select sum(number_tracks_sold) from info)*100,2)
as number_tracks_sold_perc,
genre_name 
from info 
group by genre_name
order by number_tracks_sold_perc desc

Done.


[(561, 53.38, 'Rock'),
 (130, 12.37, 'Alternative & Punk'),
 (124, 11.8, 'Metal'),
 (53, 5.04, 'R&B/Soul'),
 (36, 3.43, 'Blues'),
 (35, 3.33, 'Alternative'),
 (22, 2.09, 'Latin'),
 (22, 2.09, 'Pop'),
 (20, 1.9, 'Hip Hop/Rap'),
 (14, 1.33, 'Jazz'),
 (13, 1.24, 'Easy Listening'),
 (6, 0.57, 'Reggae'),
 (5, 0.48, 'Electronica/Dance'),
 (4, 0.38, 'Classical'),
 (3, 0.29, 'Heavy Metal'),
 (2, 0.19, 'Soundtrack'),
 (1, 0.1, 'TV Shows')]

To answer the question 
*Of the 4 albums mentioned below, we are tasked with selecting the first three albums that will be added to the store. All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:**(refer table)***



|Artist Name|Genre   |
|-----------|--------|
|Regal      |Hip-Hop |
|Red Tone   |	Punk |
|Meteor and the Girls|	Pop|
|Slim Jim Bites|	Blues|




- From what we have queried it is evident that from the 4 options given to us, Punk is the genre of music that seels the best in the United States of America which sold 130 albums which is 12.37% of the total sales.
- Blues ranks comes in at 2nd place after having sold 36 albums which is at 3.43%
- Pop and Hip-hop rank 3rd and 4th respectively after having sold 22 and 20 albums respectively

So from our analysis we will pick **Red Tone**, **Slim Jim Bites** and **Meteor and the Girls**

In [6]:
%%sql
select sum(invoice.total) as sales_dollars, 
employee.first_name||" "||employee.last_name as employee_name,
employee.country, employee.birthdate, employee.hire_date
from employee inner join customer 
on customer.support_rep_id=employee.employee_id 
inner join invoice on invoice.customer_id= customer.customer_id 
where title like "%Sales S%" group by employee.employee_id
order by employee.hire_date 

Done.


[(1731.5100000000039, 'Jane Peacock', 'Canada', '1973-08-29 00:00:00', '2017-04-01 00:00:00'),
 (1584.0000000000034, 'Margaret Park', 'Canada', '1947-09-19 00:00:00', '2017-05-03 00:00:00'),
 (1393.920000000002, 'Steve Johnson', 'Canada', '1965-03-03 00:00:00', '2017-10-17 00:00:00')]

The above query gives us the total dollar amount of sales each sales support staff made. Jane Peacock made the most sales by the dollar amount. Jane is the youngest of the 3 Sales Support Staff and had the highest dollar sales amount.
Eventhough Steve was  hired 5 months after Margaret, his sales are almost on par with hers.

In [7]:
%%sql
select customer.country,count(distinct(customer.customer_id)) as total_customers,
sum(invoice.total) as total_salves_val,
sum(invoice.total)/count(distinct(customer.customer_id)) as avg_sales_per_cus,
avg(invoice.total) as avg_ord_val
from customer
inner join invoice on customer.customer_id=invoice.customer_id
group by country
having count(distinct(customer.customer_id))>1
order by count(distinct(customer.customer_id)) desc

Done.


[('USA', 13, 1040.4899999999998, 80.0376923076923, 7.942671755725189),
 ('Canada', 8, 535.5900000000001, 66.94875000000002, 7.047236842105265),
 ('Brazil', 5, 427.68000000000006, 85.53600000000002, 7.0111475409836075),
 ('France', 5, 389.0699999999999, 77.81399999999998, 7.781399999999998),
 ('Germany', 4, 334.62, 83.655, 8.161463414634147),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Czech Republic', 2, 273.24000000000007, 136.62000000000003, 9.108000000000002),
 ('India', 2, 183.14999999999998, 91.57499999999999, 8.72142857142857),
 ('Portugal', 2, 185.13000000000002, 92.56500000000001, 6.383793103448276)]

In [8]:
%%sql
select "Other" as country, sum(total_customers) as total_customers, 
sum(total_salves_val)as total_salves_val, 
SUM(avg_sales_per_cus)/sum(total_customers) as avg_sales_per_cus,
SUM(avg_ord_val)/sum(total_customers) as avg_ord_val
from
(select customer.country as country, count(distinct(customer.customer_id)) as 
 total_customers,
sum(invoice.total) as total_salves_val,
sum(invoice.total)/count(distinct(customer.customer_id)) as avg_sales_per_cus,
avg(invoice.total) as avg_ord_val
from customer inner join invoice on 
customer.customer_id=invoice.customer_id
group by customer.country
having count(distinct(customer.customer_id))=1)

Done.


[('Other', 15, 1094.9399999999998, 72.996, 7.445071062271063)]

In [9]:
%%sql
with details_by_country as 
(
select customer.country,count(distinct(customer.customer_id)) as total_customers,
sum(invoice.total) as total_salves_val,
sum(invoice.total)/count(distinct(customer.customer_id)) as avg_sales_per_cus,
avg(invoice.total) as avg_ord_val
from customer
inner join invoice on customer.customer_id=invoice.customer_id
group by country
having count(distinct(customer.customer_id))>1
    
union 

select "Other" as country, sum(total_customers) as total_customers, 
sum(total_salves_val)as total_salves_val, 
SUM(avg_sales_per_cus)/sum(total_customers) as avg_sales_per_cus,
SUM(avg_ord_val)/sum(total_customers) as avg_ord_val
from
(select customer.country as country, count(distinct(customer.customer_id)) as 
 total_customers,
sum(invoice.total) as total_salves_val,
sum(invoice.total)/count(distinct(customer.customer_id)) as avg_sales_per_cus,
avg(invoice.total) as avg_ord_val
from customer inner join invoice on 
customer.customer_id=invoice.customer_id
group by customer.country
having count(distinct(customer.customer_id))=1)    
)

select * from details_by_country
order by total_customers

Done.


[('Czech Republic', 2, 273.24000000000007, 136.62000000000003, 9.108000000000002),
 ('India', 2, 183.14999999999998, 91.57499999999999, 8.72142857142857),
 ('Portugal', 2, 185.13000000000002, 92.56500000000001, 6.383793103448276),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Germany', 4, 334.62, 83.655, 8.161463414634147),
 ('Brazil', 5, 427.68000000000006, 85.53600000000002, 7.0111475409836075),
 ('France', 5, 389.0699999999999, 77.81399999999998, 7.781399999999998),
 ('Canada', 8, 535.5900000000001, 66.94875000000002, 7.047236842105265),
 ('USA', 13, 1040.4899999999998, 80.0376923076923, 7.942671755725189),
 ('Other', 15, 1094.9399999999998, 72.996, 7.445071062271063)]

In [10]:
%%sql
with album_track_count as 
(select invoice.invoice_id,invoice_line.invoice_line_id,invoice_line.track_id,
count(distinct(track.album_id)) as album_count
from invoice inner join invoice_line 
on invoice.invoice_id=invoice_line.invoice_id 
inner join track on track.track_id=invoice_line.track_id
group by invoice.invoice_id)

select count(*) as album_purchase from album_track_count
where album_count=1


Done.


[(171,)]

In [11]:
%%sql
with album_track_count as 
(select invoice.invoice_id,invoice_line.invoice_line_id,invoice_line.track_id,
count(distinct(track.album_id)) as album_count
from invoice inner join invoice_line 
on invoice.invoice_id=invoice_line.invoice_id 
inner join track on track.track_id=invoice_line.track_id
group by invoice.invoice_id)

select count(*) as track_purchase from album_track_count
where album_count!=1

Done.


[(443,)]

It is evident from the 2 tables above that customers who purchase individual tracks is significantly higher than customers who buy the whole album.
The question this will help us answer is as follows:

*Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.*

**Answer:** *If managment decided to do this, this would result in a signifcant loss of revenue(171 customer to be exact). I would not suggest this per the analysis made above. *

In [12]:
%%sql
select artist.name as artist_name ,count(playlist.name) as playlist_count
from playlist inner join playlist_track
on playlist.playlist_id=playlist_track.playlist_id 
inner join track on track.track_id=playlist_track.track_id
inner join album on album.album_id=track.album_id
inner join artist on artist.artist_id=album.artist_id

group by artist.name
order by count(playlist.name) desc
limit 10

Done.


[('Iron Maiden', 516),
 ('U2', 333),
 ('Metallica', 296),
 ('Led Zeppelin', 252),
 ('Deep Purple', 226),
 ('Lost', 184),
 ('Pearl Jam', 177),
 ('Eric Clapton', 145),
 ('Faith No More', 145),
 ('Lenny Kravitz', 143)]

These are the top 10 artists/bands that appear in most playlists 

In [13]:
%%sql
with purchased_or_not as
(select track.track_id as all_tracks,invoice_line.track_id as purchased_track from track left join invoice_line
on track.track_id=invoice_line.track_id)

select count(distinct(all_tracks))-count(distinct(purchased_track)) as unpurchased_tracks,
count(distinct(purchased_track)) as purchased_tracks
from purchased_or_not 

Done.


[(1697, 1806)]

In [14]:
%%sql
select count(track_id) from track

Done.


[(3503,)]

In [15]:
%%sql
select count(distinct(track_id)) from invoice_line

Done.


[(1806,)]

3503 tracks are in the store and 1806 are the number of distinct tracks purchased by the customers.

In [16]:
%%sql 
create view purchased_tracks as 
select count(distinct(invoice_line.track_id)) as tracks_by_genre_purchased,
genre.name as genre_name
from invoice_line inner join track on invoice_line.track_id=track.track_id
inner join genre on track.genre_id=genre.genre_id
group by genre.genre_id

(sqlite3.OperationalError) table purchased_tracks already exists
[SQL: create view purchased_tracks as 
select count(distinct(invoice_line.track_id)) as tracks_by_genre_purchased,
genre.name as genre_name
from invoice_line inner join track on invoice_line.track_id=track.track_id
inner join genre on track.genre_id=genre.genre_id
group by genre.genre_id]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


These are the number of tracks purchased by customers. 
Now let us look at tracks by genre present in the store.

In [17]:
%%sql 
create view all_tracks as
select count(distinct(track.track_id)) number_of_tracks, 
genre.name as genre_name from track 
inner join genre on track.genre_id=genre.genre_id
group by genre.name 

(sqlite3.OperationalError) table all_tracks already exists
[SQL: create view all_tracks as
select count(distinct(track.track_id)) number_of_tracks, 
genre.name as genre_name from track 
inner join genre on track.genre_id=genre.genre_id
group by genre.name]
(Background on this error at: http://sqlalche.me/e/13/e3q8)


In [18]:
%%sql 
select all_tracks.genre_name, all_tracks.number_of_tracks as total_tracks_in_store,
purchased_tracks.tracks_by_genre_purchased as tracks_purchased,
round(cast(purchased_tracks.tracks_by_genre_purchased as float)/all_tracks.number_of_tracks*100,2) as percentage_unsold
from all_tracks inner join purchased_tracks 
on all_tracks.genre_name=purchased_tracks.genre_name
order by percentage_unsold

Done.


[('Drama', 64, 1, 1.56),
 ('TV Shows', 93, 2, 2.15),
 ('Soundtrack', 43, 5, 11.63),
 ('Latin', 579, 119, 20.55),
 ('Classical', 74, 16, 21.62),
 ('Heavy Metal', 28, 7, 25.0),
 ('Reggae', 58, 22, 37.93),
 ('Jazz', 130, 61, 46.92),
 ('Pop', 48, 25, 52.08),
 ('Alternative & Punk', 332, 176, 53.01),
 ('Hip Hop/Rap', 35, 21, 60.0),
 ('Metal', 374, 238, 63.64),
 ('Blues', 81, 56, 69.14),
 ('Rock', 1297, 915, 70.55),
 ('Alternative', 40, 34, 85.0),
 ('R&B/Soul', 61, 55, 90.16),
 ('Electronica/Dance', 30, 29, 96.67),
 ('Easy Listening', 24, 24, 100.0)]

- The table above very clearly informs us that tracks in the Drama and TV Shows genre almost go unsold. So if managemnet was looking to remove tracks of any particular genre these 2 would be the best. 
- In terms of number Rock, Metal, Alternative & Punk ans Latin sell the most. 
- In terms of highest percentage of tracks sold, Easy Listening Electronica/Dance, R&B Soul sells the best, they almost sell all the copies

In [19]:
%%sql

select 
"Protected" as type,
count(track.track_id) as count_tracks
from media_type inner join track 
on media_type.media_type_id=track.media_type_id
where media_type.name like "%Protected%"

union 
select "Non-Protected" as type, 
count(track.track_id) from media_type 
inner join track on media_type.media_type_id=track.media_type_id 
where media_type.name not like "%Protected%"

Done.


[('Non-Protected', 3052), ('Protected', 451)]

Most of the tracks falls under the Non-Protected category

# Conclusion
Let us summarize what we have done in this project.

1. Based on US sales data, management at Chinook should sign these artists:
    - Red Tone
    - Meteor and the Girls
    - Slim Jim Bites
2. If management go ahead with their strategy of buying only the most popular tracks from every album it would result in a signifcant loss of revenue(171 customer to be exact), but if they were to do this, It would definetely lead some loss in income in the short run and hopefully a positive return in the future.
3. Over half of the tracks do not sell
4. The 2 Northern American  countries(USA and Canada) are the highest contributors with regards to income and customer presence to Chinook. 
5. Iron Maiden,U2 and Metallica and Led Zepplin are the artists found in most playlists. These are Rock and Heavy Metal bands/artists,perfectly paving a segway into our next point conclusion.
6. Rock and Heavy Metal as genres sold the most tracks, and the Drama and TV Shows genres sold almost none of their tracks.
7. The media type being protected does not influence the customers into buying it, as majority of buys happen to be of the Non-protected types
